intermediate_source/parametrizations.py

```diff
@@ -0,0 +1,353 @@
+symmetric = Symmetric()
+assert torch.allclose(layer.weight, symmetric(layer.parametrizations.weight.original))
+
+###############################################################################
+# Parametrizations are first-class citizens
+# -----------------------------------------
+# Since ``layer.parametrizations`` is an `nn.ModuleList`, it means that the parametrizations
+# are properly registered as submodules of the original module. As such, the same rules
+# for registering parameters in a module apply to register a parametrization.
+# For example, if a parametrization has parameters, these will be moved from CPU
+# to CUDA when calling ``model = model.cuda()``.
+#
+# Caching the value of a parametrization
+# --------------------------------------
+# Parametrizations come with an inbuilt caching system via the context manager ``P.cached()``
+class NoisyParametrization(nn.Module):
+    def forward(self, X):
+        print("Computing the Parametrization")
+        return X
+
+layer = nn.Linear(2, 3)
+P.register_parametrization(layer, "weight", NoisyParametrization())
+print("Here, layer.weight is recomputed every time we call it")
+foo = layer.weight + layer.weight.T
+bar = layer.weight.sum()
+with P.cached():
+    print("Here, it is computed just the first time layer.weight is called")
+    foo = layer.weight + layer.weight.T
+    bar = layer.weight.sum()
+
+###############################################################################
+# Composing Parametrizations
+# --------------------------
+# Concatenating two parametrizations is as easy as registering them on the same tensor.
+# We may use this to create complex parametrizations from simple ones. For example, the
+# `Cayley map <https://en.wikipedia.org/wiki/Cayley_transform#Matrix_map>`_
+# maps the skew-symmetric matrices to the orthogonal matrices of positive determinant. We can
+# concatenate ``Skew`` and a parametrization that implements the Cayley map to get a layer with
+# orthogonal weight
+class CayleyMap(nn.Module):
+    def __init__(self, n):
+        self.register_buffer("Id", torch.eye(n))
+
+    def forward(self, X):
+        # (I + X)(I - X)^{-1}
+        return torch.solve(self.Id + X, self.Id - X).solution
+
+layer = nn.Linear(3, 3)
+P.register_parametrization(layer, "weight", Skew())
+P.register_parametrization(layer, "weight", CayleyMap(3))
+X = layer.weight
+assert torch.allclose(X.T @ X, torch.eye(3))  # X is orthogonal
+
+###############################################################################
+# This may also be used to prune a parametrized module, or to reuse parametrizations. For example,
+# we may use the fact that the exponential of matrices maps the symmetric matrices to the
+# Symmetric Positive Definite (SPD) matrices, and the skew-symmetric matrices to the orthogonal
+# matrices. Using these two facts, we may reuse the parametrizations
+class MatrixExponential(nn.Module):
+    def forward(X):
+        return torch.matrix_exp(X)
```

Comment:
Could `torch.matrix_exp` be directly used instead of `MatrixExponential()`?

Reply:
No, it could not. At the moment `Parametrizations` are defined to be `nn.Modules`, so they do not support the functional API.

I do not know whether it would be necessary to discuss this, as in the whole tutorial it's been made clear that a parametrization is just a plain `nn.Module`. What do you think @albanD ?

Reply:
My reasoning: in many places in PyTorch one expects both functions and module objects work (especially in older areas of PyTorch). Whenever this is violated - quantization, module tracing etc, I'm always suspicious of undeclared magic :)

In this case, it begs the question, since the module here is just a wrapper

Reply:
That is indeed an interesting question.
But for now everything is expected to be a nn.Module and that sounds enough for me.
If we want to relax that in the future we might be able to do so but that would be out of scope of this tutorial.